File: src/main/java/lk/tech/commands/impl/AbstractCommand.java
package lk.tech.commands.impl;

import lk.tech.dto.Answer;
import lombok.Getter;

@Getter
public abstract class AbstractCommand {

    private AbstractCommand next;

    public Answer analyze(String command) {
        if (command.equals(condition())) {
            return run(command);
        }
        AbstractCommand next = getNext();
        if (next != null) {
            return next.analyze(command);
        }
        return new Answer("/unknown", null);
    }


    abstract Answer run(String command);

    abstract String condition();

    public AbstractCommand setNext(AbstractCommand next) {
        this.next = next;
        return next;
    }
}

File: services/commands/abstract_command.go
package commands

import (
	"awesomeProject/dto"
)

// AbstractCommand — интерфейс как в Java (analyze, run, condition)
type AbstractCommand interface {
	Analyze(cmd string) dto.ResultInterface
	Run(cmd string) dto.ResultInterface
	Condition() string
	SetNext(next AbstractCommand)
	GetNext() AbstractCommand
}

// BaseCommand — базовая структура, аналог абстрактного класса в Java.
// Хранит "next" и общую логику Analyze().
type BaseCommand struct {
	next AbstractCommand
}

// SetNext — установка следующей команды
func (bc *BaseCommand) SetNext(next AbstractCommand) {
	bc.next = next
}

// GetNext — возврат next
func (bc *BaseCommand) GetNext() AbstractCommand {
	return bc.next
}

// Analyze — общая логика как в Java (chain of responsibility)
func (bc *BaseCommand) Analyze(cmd string, self AbstractCommand) dto.ResultInterface {
	if cmd == self.Condition() {
		return self.Run(cmd)
	}

	if bc.next != nil {
		return bc.next.Analyze(cmd)
	}

	return dto.Result{Status: "Unknown"}
}
